R/create_single_dose_dataset.R

```diff
@@ -0,0 +1,378 @@
+                                       dose_freq = EXDOSFRQ,
+                                       start_date = ASTDT,
+                                       end_date = AENDT,
+                                       lookup_table = dose_freq_lookup,
+                                       lookup_column = CDISCSubmissionValue) {
+
+  col_names <- colnames(dataset)
+  dose_freq <- assert_symbol(enquo(dose_freq))
+  lookup_column <- assert_symbol(enquo(lookup_column))
+  start_date <- assert_symbol(enquo(start_date))
+  end_date <- assert_symbol(enquo(end_date))
+  assert_data_frame(dataset, required_vars = quo_c(dose_freq, start_date, end_date))
+  assert_data_frame(lookup_table, required_vars = vars(DoseWindow, DoseCount, ConversionFactor))
+
+    # Set up lookup table to be joined to dataset
+
+    lookup <- lookup_table %>%
+      rename(!!dose_freq := !!lookup_column)
+
+    # Check values of lookup vs. data and return error if values are not covered
+
+    value_check <- dataset %>%
+      select(!!dose_freq) %>%
+      anti_join(lookup, by = as.character(quo_get_expr(dose_freq))) %>%
+      unique()
+
+    if (nrow(value_check) > 0) {
+      values_not_found <- str_c(value_check %>% select(!!dose_freq), collapse = ", ")
+
+      err_msg <- paste0(
+        sprintf("The following values of %s in %s do not appear in %s:\\n",
+                as.character(quo_get_expr(dose_freq)),
+                arg_name(substitute(dataset)),
+                arg_name(substitute(lookup_table))
+                ), values_not_found)
+    }
+
+    # Use compute_duration to determine the number of completed dose periods
+    dataset_part_1 <- dataset %>%
+      filter(!!dose_freq == "ONCE")
+
+    dataset_part_2 <- dataset %>%
+      filter(!!dose_freq != "ONCE")
+
+    dataset_part_2 <- dataset_part_2 %>%
+      left_join(lookup, by = as.character(quo_get_expr(dose_freq))) %>%
+      mutate(dose_periods = case_when(
+          DoseWindow == "MINUTE" ~ compute_duration(!!start_date, !!end_date, out_unit = "minutes"),
+          DoseWindow == "HOUR" ~ compute_duration(!!start_date, !!end_date, out_unit = "hours"),
+          DoseWindow == "DAY" ~ compute_duration(!!start_date, !!end_date, out_unit = "days"),
+          DoseWindow == "WEEK" ~ compute_duration(!!start_date, !!end_date, out_unit = "weeks"),
+          DoseWindow == "MONTH" ~ compute_duration(!!start_date, !!end_date, out_unit = "months"),
+          DoseWindow == "YEAR" ~ compute_duration(!!start_date, !!end_date, out_unit = "years")
+      )
+        ) %>%
+      mutate(dose_count = ceiling(dose_periods * DoseCount)) %>%
+      derive_var_obs_number(new_var = grpseq)
+
+    # Flag to determine if date or datetime must be returned
+
+    time_flag <- nrow(dataset_part_2 %>%
+                         filter(DoseWindow %in% c("MINUTE", "HOUR"))) > 0
+
+    if (time_flag) {
+      dataset_part_1 <- dataset_part_1 %>%
+        mutate(!!start_date := as_datetime(!!start_date),
```

Comment:
I think the `start_date` and `end_date` column should not be changed.

Somehow this seems to create issues.

For example the following example fails.
```
tex <- tibble::tribble(
    ~USUBJID, ~ASTDTM, ~AENDTM, ~EXDOSFRQ, ~EXTRT, ~temp_obs_nr,
    "1", ymd_hms("2020-02-07T02:00:00"), ymd_hms("2020-02-09T01:00:00"), "QD", "DRUG 1", 1,
    "1", ymd_hms("2020-02-07T02:00:00"), ymd_hms("2020-02-09T01:00:00"), "QD", "DRUG 2", 2
) %>%
    mutate(ASTDT = date(ASTDTM),
           AENDT = date(AENDTM))

create_single_dose_dataset(tex,
                           start_date = ASTDTM,
                           end_date = AENDTM)
```

Reply:
i think this fails because currently the `time_flag` is set based on the explicit presence of a dose frequency like HOUR or MINUTE in the data, so it's trying to turn ASTDTM/AENDTM into regular date objects.

it seems to me like the function might need to return dates for `start_date`/`end_date` in some cases and datetimes for the same variables in other cases. `time_flag` was my attempt at determining when to do this, but here were my other considerations:

- `start_time` and `end_time` in addition to `start_date` and `end_date`
- expand the conditions for when `time_flag` is positive to see if the arguments given to `start_date` and `end_date` are datetimes

this second one seems more in line with your suggestion. i think i'd still need to change the type of the column in a case like below, where a date is fed as `start_date`/`end_date` but the frequency involves hours.

```
tex <- tibble::tribble(
    ~USUBJID, ~ASTDTM, ~AENDTM, ~EXDOSFRQ, ~EXTRT, ~temp_obs_nr,
    "1", ymd_hms("2020-02-07T02:00:00"), ymd_hms("2020-02-09T01:00:00"), "Q4H", "DRUG 1", 1,
    "1", ymd_hms("2020-02-07T02:00:00"), ymd_hms("2020-02-09T01:00:00"), "Q4H", "DRUG 2", 2
) %>%
    mutate(ASTDT = date(ASTDTM),
           AENDT = date(AENDTM))

create_single_dose_dataset(tex,
                           start_date = ASTDT,
                           end_date = AENDT)
```

Reply:
We should not change the type of the input variables. If the type of `ASTDT` is changed from date to datetime, it is quite confusing and not ADaM compliant.

I think we have two options:

- We expect a datetime variable for `start_date` and `end_date` and stop if another type of variable is provided.
- We accept date and datetime variables for `start_date` and `end_date`. In the function we create a temporary datetime variable for the computation and convert it to a date variable at the end (if necessary). If a date variable is provided and the computation uses hours or minutes, we could issue a warning or error.

Reply:
sounds good. i will implement the second option. thanks for your patience and guidance!